include/soci/callbacks.h

```diff
@@ -0,0 +1,47 @@
+//
+// Copyright (C) 2015 Maciej Sobczak
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef SOCI_CALLBACKS_H_INCLUDED
+#define SOCI_CALLBACKS_H_INCLUDED
+
+namespace soci
+{
+
+class session;
+
+// Simple callback interface for reporting failover events.
+// The meaning of each operation is intended to be portable,
+// but the behaviour details and parameters can be backend-specific.
+class SOCI_DECL failover_callback
+{
+public:
+
+    // Called when the failover operation has started,
+    // after discovering connectivity problems.
+    virtual void started() {}
```

Comment:
It looks like this and other methods are always called inside `try/catch(...)`, so IMHO it would make sense to make them protected and have public `call_started()` and so on which would call the real callback and eat the exceptions from it.

This would reduce the amount of `try/catch`s in the rest of the code and ensure that they can't be forgotten, letting the exceptions escape.